### ISSUE TITLE
Server does not work with tagXplorer custom key ?

### ISSUE THREAD

**Issue by Nyl000** [icedevml/ntag424-backend#34]
Hello, 

My tags work well with this project by using the factory key 00000000000000000000000000000000 but when I try to change the key via tagXplorer.

![2022-10-11 02_14_06-Window](https://user-images.githubusercontent.com/857787/194971173-db2fcdb1-2887-48f1-a6f0-1fcf58d5c53c.png)
+ 
`
docker run \
    -p 5000:80 \
    -e SDM_MASTER_KEY=00000000000000000000000000000001 \
    icedevml/sdm-backend:build20220822-v2
`
It doesn't work. 

When I try with the demo mobile app with key 00000000000000000000000000000001, it works. 

Am I missing something ? 

Thanks and have a nice day.

**Comment by icedevml** [comment]
The `SDM_MASTER_KEY` is not the key which is used directly with the tags. There is a derivation algorithm involved to transform `SDM_MASTER_KEY` into the tag-specific key. This algorithm is present here:
https://github.com/icedevml/sdm-backend/blob/4d75ab6ac6dc77d0988f236028194acef016fe6c/derive.py

Sorry for late reply, I didn't got any notification out of this unfortunately.

**Comment by CerebrallyPerplexed** [comment]
I have a similar issue...

NTAG424 DNA KEY0: 00000000000000000000000000000000
KEYS 1-4 ALSO SAME 00000000000000000000000000000000
SDM_MASTER_KEY: 00000000000000000000000000000000

All working fine via SDM

But after changing KEY0 on tag > SDM Still working fine... (but shouldn't as master-key changed on tag?)
If I change SDM_MASTER_KEY in config.py (or is it config.dist.py? did both anyway) to the NEW KEY0 value, > Bad Request, probably wrong signature

Is this because SDM Access Rights are set to 01 & 02? I.e. both still set to 00000000000000000000000000000000 ?

Ideally, I would wish to change ALL tag KEYS0-4 to my own keys - but will this kill the SDM?

Edit: I have changed the Read Access Key, since adding screenshot below, to 0E as this ensures that the NDEF record can be read without requiring authentication (on a mobile etc.) or so I believe?

![image](https://user-images.githubusercontent.com/127226654/223473877-e189d95b-8844-4e9a-9add-43d2455010d2.png)

**Comment by icedevml** [comment]
Hello,

the `SDM_MASTER_KEY` is not directly applied to the tag. Instead, it's plugged to these functions:
https://github.com/icedevml/sdm-backend/blob/4d75ab6ac6dc77d0988f236028194acef016fe6c/derive.py

You need to use the following script in order to convert `SDM_MASTER_KEY` into a set of keys that has to be configured on the tag:
```python
import binascii
# store this file as "derive.py": https://github.com/icedevml/sdm-backend/blob/4d75ab6ac6dc77d0988f236028194acef016fe6c/derive.py
from derive import derive_tag_key, derive_undiversified_key

SDM_MASTER_KEY = binascii.unhexlify("00000000000000000000000000000001")
TAG_UID = binascii.unhexlify("... put your tag's UID here...")

master_key = derive_tag_key(SDM_MASTER_KEY, TAG_UID, 0)
key_1 = derive_undiversified_key(SDM_MASTER_KEY, 1)
key_2 = derive_tag_key(SDM_MASTER_KEY, TAG_UID, 2)
key_3 = derive_tag_key(SDM_MASTER_KEY, TAG_UID, 3)
key_4 = derive_tag_key(SDM_MASTER_KEY, TAG_UID, 4)

print('key 0', master_key.hex())
print('key 1', key_1.hex())
print('key 2', key_2.hex())
print('key 3', key_3.hex())
print('key 4', key_4.hex())
```

Hope that this helps.

**Comment by lukagrbec** [comment]
Hello, I've been trying to implement the code you suggested to get the custom keys, but I'm not sure where to go from here to make it work. Should I compare the key values with the ones on the tag (I'm don't know how to get their value) or should they be used in a modified version of `decrypt_sun_message` function (which takes in keys as parameters?)? I would sincerely appreciate any further guidance.

**Comment by icedevml** [comment]
@lukagrbec You need to launch this script in order to generate tag-individual keys out of the `SDM_MASTER_KEY`. This script will print out 5 keys numbered from 0 to 4.

You would need to take the factory-clean NFC tag and set these five keys to the values indicated by the script mentioned above. Also, you would need to configure the dynamic URL correctly and then the tag will be correctly accepted by sdm-backend.

**Comment by lukagrbec** [comment]
Thank you for your reply. I managed to get the 5 keys and configured them on the tag. However, I can't get the sdm-backend to work as expected. When I scan the tag it raises "Unsupported UID length" in decrypt_sun_message function in libsdm.py. 

I changed the sdm_master_key to 00000000000000000000000000000001 in config.py and config.dist.py. I used the same sdm_master_key in the script to generate the custom keys. For the tag_uid="04577252B97380" I got the following keys:
key 0 072f088a9aeab132824c597c60225e06
key 1 84efbccbe0463137112231a49ee69a8a
key 2 aa35cb33fa140e861b7bcc1f9ce33382
key 3 fcb413ea55d301a6e094f9144d70add5
key 4 8bcaad09c8b0004131439db438ea9645

Is there anything else I should change in the sdm-backend?

**Comment by icedevml** [comment]
@lukagrbec Which version of sdm-backend are you using? I recommend to try to downgrade to:
https://github.com/icedevml/sdm-backend/tree/a89a8381a7b680abff721f006085ec4d15f8c543
and try again.

The "unsupported UID length" means that the decryption of PICCData has failed at the very early stage. So something is wrong around the keys.

**Comment by lukagrbec** [comment]
@icedevml I downgraded and I still get the error that UID length is unsupported, however, if I print the uid_length now it returns 10, whereas previously it returned 0 (this is the tag's uid 04577252B97380).

Should I after setting custom keys also set the url again in TagWriter? I first set the url and then changed the keys in TagExplorer.

**Comment by icedevml** [comment]
@lukagrbec There is no need to rewrite the URL after changing the keys.

The "UID length is unsupported" error is due to the fact that PICCData decryption has failed. Please check around the key 1, it looks like the value expected by sdm-backend is different from the value that you've configured on the tag.

I can confirm that the keys that you've pasted are generated correctly for tag UID `04577252B97380` and `SDM_MASTER_KEY` = `00000000000000000000000000000001`.

Could you check if your `SDM_MASTER_KEY` is actually set to `00000000000000000000000000000001` on the sdm-backend and that the service was restarted? Could you also open TagXplorer and try to authenticate with these keys?

```
key 0 072f088a9aeab132824c597c60225e06
key 1 84efbccbe0463137112231a49ee69a8a
key 2 aa35cb33fa140e861b7bcc1f9ce33382
```

Basically it sounds like at least the key 1 is not configured correctly on the tag, and that's why validation doesnt pass.

**Comment by lukagrbec** [comment]
Thanks again for your reply. I opened TagXplorer and autheticated the keys. I also printed SDM_MASTER_KEY in app.py and get this value: `b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01'`. 
I also printed sdm_meta_read_key: `b'\x84\xef\xbc\xcb\xe0F17\x11"1\xa4\x9e\xe6\x9a\x8a'`. (it's similar to key1. Should it be identical?). (Sdm_meta_read_key is derived from `derive_undiversified_key(SDM_MASTER_KEY,` 1)`).

**Comment by icedevml** [comment]
Hello @lukagrbec,

these are the same:
```
b'\x84\xef\xbc\xcb\xe0F17\x11"1\xa4\x9e\xe6\x9a\x8a'
84efbccbe0463137112231a49ee69a8a
```

just in different representation.

Could you authenticate with key 1 in TagXplorer? Please paste the second representation into TagXplorer.

**Comment by lukagrbec** [comment]
Here is a screenshot from TagXplorer. I clicked Authenticate first and it opened Authenticity verification successful. Is there anything else I should do?
![authentication](https://user-images.githubusercontent.com/63504896/229225379-e46884f3-3f6d-4b69-83fa-139077e97058.PNG)

**Comment by lukagrbec** [comment]
Hi @icedevml, I've managed to make it work! I had to change the SDM Meta Read Key. Thank you so much for the help.

**Comment by icedevml** [comment]
@lukagrbec Great, thanks for letting me know! I will close this as resolved.

**Comment by enr0090** [comment]
> Hi [@icedevml](https://github.com/icedevml), I've managed to make it work! I had to change the SDM Meta Read Key. Thank you so much for the help.

Hi @lukagrbec I know this is an old thread, but I'm facing the exact same issue.
I don't think the backend is the problem, since I'm able to reproduce the results you're getting using the UID `04577252B97380`.
Would it be possible to share your TagXplorer configuration somehow—maybe via screenshots? That would be a huge help. Thanks!

**Comment by miag676** [comment]
@enr0090 It's not the backend as you wrote. You need to go into File Management and then in SDM Access Writes change SDM Meta Read Key (the three values from 00 to 01, 02 and 01). I attached the screenshot. 

<img width="1919" height="1010" alt="Image" src="https://github.com/user-attachments/assets/d9e5ff37-b3a2-4df7-9855-7dbbe6d2842e" />
